File: web/src/store/market.ts
// web/src/store/market.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import { marketApi } from '@/api/market'
import type { WritingMemory } from '@/api/memory'
import type { MemoryReview } from '@/api/market'

export const useMarketStore = defineStore('market', () => {
  const memories = ref<WritingMemory[]>([])
  const total = ref(0)
  const currentMemory = ref<Record<string, any> | null>(null)
  const reviews = ref<MemoryReview[]>([])
  const reviewTotal = ref(0)
  const loading = ref(false)

  async function fetchMemories(params?: { category?: string; keyword?: string; order_by?: string; page?: number; page_size?: number }) {
    loading.value = true
    try {
      const data: any = await marketApi.listMemories(params)
      memories.value = Array.isArray(data?.list) ? data.list : []
      total.value = data?.total || 0
    } finally {
      loading.value = false
    }
  }

  async function fetchMemory(mid: number) {
    const data: any = await marketApi.getMemory(mid)
    currentMemory.value = data || null
    return data
  }

  async function buyMemory(mid: number) {
    return await marketApi.buy(mid)
  }

  async function fetchReviews(mid: number, page = 1, pageSize = 20) {
    const data: any = await marketApi.listReviews(mid, { page, page_size: pageSize })
    reviews.value = Array.isArray(data?.list) ? data.list : []
    reviewTotal.value = data?.total || 0
  }

  async function submitReview(mid: number, rating: number, comment?: string) {
    await marketApi.submitReview(mid, { rating, comment })
    await fetchReviews(mid)
  }

  function reset() {
    memories.value = []
    total.value = 0
    currentMemory.value = null
    reviews.value = []
    reviewTotal.value = 0
  }

  return {
    memories,
    total,
    currentMemory,
    reviews,
    reviewTotal,
    loading,
    fetchMemories,
    fetchMemory,
    buyMemory,
    fetchReviews,
    submitReview,
    reset,
  }
})

File: server/internal/handler/market.go
// server/internal/handler/market.go
package handler

import (
	"net/http"
	"strconv"

	"ai-curton/server/internal/service"

	"github.com/gin-gonic/gin"
)

// MarketHandler 市场请求处理器
type MarketHandler struct {
	marketSvc *service.MarketService
}

// NewMarketHandler 创建 MarketHandler 实例
func NewMarketHandler(marketSvc *service.MarketService) *MarketHandler {
	return &MarketHandler{marketSvc: marketSvc}
}

// ListMemories 市场浏览
// GET /api/v1/market/memories
func (h *MarketHandler) ListMemories(c *gin.Context) {
	category := c.Query("category")
	keyword := c.Query("keyword")
	orderBy := c.DefaultQuery("order_by", "sales")
	page, _ := strconv.Atoi(c.DefaultQuery("page", "1"))
	pageSize, _ := strconv.Atoi(c.DefaultQuery("page_size", "20"))
	genreID, _ := strconv.ParseUint(c.Query("genre_id"), 10, 64)

	memories, total, err := h.marketSvc.ListMarketMemories(category, keyword, orderBy, page, pageSize, uint(genreID))
	if err != nil {
		InternalError(c, err.Error())
		return
	}

	Success(c, gin.H{
		"list":  memories,
		"total": total,
		"page":  page,
	})
}

// GetMemory 市场记忆详情
// GET /api/v1/market/memories/:mid
func (h *MarketHandler) GetMemory(c *gin.Context) {
	mid, err := strconv.ParseUint(c.Param("mid"), 10, 64)
	if err != nil {
		BadRequest(c, "invalid memory id")
		return
	}

	userID := c.GetUint("user_id")
	result, err := h.marketSvc.GetMarketMemory(uint(mid), userID)
	if err != nil {
		Error(c, http.StatusNotFound, "memory not found")
		return
	}

	Success(c, result)
}

// Buy 购买记忆
// POST /api/v1/market/memories/:mid/buy
func (h *MarketHandler) Buy(c *gin.Context) {
	mid, err := strconv.ParseUint(c.Param("mid"), 10, 64)
	if err != nil {
		BadRequest(c, "invalid memory id")
		return
	}

	userID := c.GetUint("user_id")
	order, err := h.marketSvc.PurchaseMemory(userID, uint(mid))
	if err != nil {
		BadRequest(c, err.Error())
		return
	}

	Success(c, order)
}

// ListReviews 评价列表
// GET /api/v1/market/memories/:mid/reviews
func (h *MarketHandler) ListReviews(c *gin.Context) {
	mid, err := strconv.ParseUint(c.Param("mid"), 10, 64)
	if err != nil {
		BadRequest(c, "invalid memory id")
		return
	}

	page, _ := strconv.Atoi(c.DefaultQuery("page", "1"))
	pageSize, _ := strconv.Atoi(c.DefaultQuery("page_size", "20"))

	reviews, total, err := h.marketSvc.ListReviews(uint(mid), page, pageSize)
	if err != nil {
		InternalError(c, err.Error())
		return
	}

	Success(c, gin.H{
		"list":  reviews,
		"total": total,
		"page":  page,
	})
}

// SubmitReview 提交评价
// POST /api/v1/market/memories/:mid/reviews
func (h *MarketHandler) SubmitReview(c *gin.Context) {
	mid, err := strconv.ParseUint(c.Param("mid"), 10, 64)
	if err != nil {
		BadRequest(c, "invalid memory id")
		return
	}

	userID := c.GetUint("user_id")

	var req struct {
		Rating  int    `json:"rating" binding:"required,min=1,max=5"`
		Comment string `json:"comment"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		BadRequest(c, err.Error())
		return
	}

	if err := h.marketSvc.SubmitReview(userID, uint(mid), req.Rating, req.Comment); err != nil {
		BadRequest(c, err.Error())
		return
	}

	SuccessWithMessage(c, "review submitted", nil)
}
